zap/src/main/java/org/zaproxy/zap/extension/brk/ExtensionBreak.java

```diff
@@ -211,19 +215,103 @@ private BreakpointsPanel getBreakpointsPanel() {
         return breakpointsPanel;
     }
 
+    private void readAllBreakpointsFromDB() {
+        TableBreak tableBreak = getModel().getDb().getTableBreak();
+        try {
+            Vector<Integer> breakIdsList = tableBreak.getBreakList();
+            for (Integer brkId : breakIdsList) {
+                RecordBreak recordBreak = tableBreak.read(brkId);
+                HttpBreakpointMessage httpBreakpoint =
+                        new HttpBreakpointMessage(
+                                recordBreak.getUrlString(),
+                                Location.valueOf(recordBreak.getLocation()),
+                                Match.valueOf(recordBreak.getMatch()),
+                                recordBreak.isInverse(),
+                                recordBreak.isIgnoreCase(),
+                                recordBreak.isOnRequest(),
+                                recordBreak.isOnResponse());
+                httpBreakpoint.setBrkId(brkId);
+                this.getBreakpointsPanel().addBreakpoint(httpBreakpoint);
+            }
+        } catch (DatabaseException e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+    }
+
     public void addBreakpoint(BreakpointMessageInterface breakpoint) {
         this.getBreakpointsPanel().addBreakpoint(breakpoint);
         // Switch to the panel for some visual feedback
         this.getBreakpointsPanel().setTabFocus();
+        writeBreakToDB(breakpoint);
+    }
+
+    private void writeBreakToDB(BreakpointMessageInterface breakpoint) {
+        HttpBreakpointMessage httpBreakpoint = (HttpBreakpointMessage) breakpoint;
```

Comment:
What happens with other breakpoint types? e.g. WebSocket

Reply:
I got confused at first. HttpBreakpointMessage is the only implementation in core.
Extensions should have their own implementations to work with DB.
So I add instance class checks before methods call.